### PR TITLE
added initial semantic segmentation example

### LINKED ISSUE #635: Semantic Segmentation example
## 🚀 Feature

Semantic Segmentation example code with dataloading and training implemented.
### Motivation

There are not many examples available for PyTorch Lightning (PL) as of now. A reproducible example illustrating semantic segmentation will be helpful for users to understand how everything works. I had to look around a lot while using PL, i.e. the documentation is not sufficient yet. So, this will improve the situation.
### Pitch

I have already implemented semantic segmentation using PL in my repository [here](https://github.com/akshaykvnit/pl-sem-seg) using the KITTI dataset and a few models (like ResNet50 and 101 FCN, DeepLabv3 - these are available directly in `torchvision`). So, the example will be similar i.e. it will illustrate the dataloading, training step and optimizer configuration steps clearly.

I can open a pull request implementing this if the idea is acceptable.

### Alternatives

The existing examples are too complicated (even the basic ones have way too much functionality mentioned for someone starting out with PL). The basic one is based on MNIST classification while the domain one is on GAN, so this can be an addition to the domain-specific examples.

### Additional context

Check [my repo](https://github.com/akshaykvnit/pl-sem-seg). It took me a lot of time to implement this simple thing :grimacing: and wouldn't have been possible using only the pl-examples in the repo (I had to look around a lot).

### PR DESCRIPTION
**Pull request by akshaykvnit**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?   
- [ ] Did you write any new necessary tests?  

## What does this PR do?
Fixes #635 (Semantic Segmentation example).

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Some additional info
I tried to add a `test_step` in the `LightningModule` which logs the predicted segmentation output. Since the default logger is TensorBoard, I expected the images to show up when I use `tensorboard logdir=lightning_logs` (at the default save path). But, they don't show up. I couldn't find any example which uses the logger properly. So, if someone can point me to an example, then I'll finish that (right now it's commented out).

Also, I have kind of followed the GAN example as a template, so I have put all of the code in a single  file. If you prefer, then I can separate the `Dataset` class into another file. And I have added this as a `full_example`, so please let me know if you want it in `domain_examples` or something.

Also, I have used the [KITTI dataset](http://www.cvlibs.net/datasets/kitti/) (since it is small and enough for an example) but it is not available directly in `torchvision`. So, it needs to be downloaded separately and the path needs to be given in the `hparams`. Note that the segmentation datasets available in torchvision (like Pascal VOC) are too large for a simple example. 

Also, I couldn't figure out what documentation or tests are required, because none of the other examples have any documentation or tests (that I could find). Please point out if anything is required.

### CODE REVIEW

**Comment by williamFalcon** [comment]
@akshaykvnit awesome. can you get the tests to pass and update the docs?

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    )
+
+    # ------------------------
+    # 3 START TRAINING
+    # ------------------------
+    trainer.fit(model)
+
+if __name__ == '__main__':
+    parser = ArgumentParser()
+    parser.add_argument("--root", type = str, help = "path where dataset is stored")
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: I guess lint does not like spaces around =

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    # 2 INIT TRAINER
+    # ------------------------
+    trainer = pl.Trainer(
+        gpus = hparams.gpus
+    )
+
+    # ------------------------
+    # 3 START TRAINING
+    # ------------------------
+    trainer.fit(model)
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: It would be nice to also add inference usage, like segment an image from dataset...

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+import torchvision
+import torchvision.transforms as transforms
+from torch.utils.data import DataLoader, Dataset
+from torchvision.models.segmentation import fcn_resnet50
+
+from PIL import Image
+import pytorch_lightning as pl
+
+class KITTI(Dataset):
+    def __init__(self, root_path, split = 'test', img_size = (1242, 376), transform = None):
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: flake does not like the spaces around =

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        Sets void classes to zero so they won't be considered for training
+        '''
+        for voidc in self.void_labels :
+            mask[mask == voidc] = self.ignore_index
+        for validc in self.valid_labels :
+            mask[mask == validc] = self.class_map[validc]
+        return mask
+    
+    def get_filenames(self, path):
+        files_list = list()
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: pls add docstring

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+        for filename in os.listdir(path):
+            files_list.append(os.path.join(path, filename))
+        return files_list
+    
+class SegModel(pl.LightningModule):
+    def __init__(self, hparams):
+        super(SegModel, self).__init__()
+        self.hparams = hparams
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: do you need to keep it since you assign all other parameters...

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    
+    def training_step(self, batch, batch_nb) :
+        img, mask = batch
+        img = img.float()
+        mask = mask.long()
+        out = self.forward(img)
+        loss_val = F.cross_entropy(out['out'], mask, ignore_index = 250)
+        return {'loss' : loss_val}
+    
+#     def test_step(self, batch, batch_nb):
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: why is this commented? if it not needed, pls remove it...

**Comment by Borda** [comment]
@williamFalcon a bit more general thought is it would be better to write examples as ipython notebooks which we can transform to nice pages in Docs even with images as I did [here](https://pyimsegm.readthedocs.io/en/latest/notebooks/segment-2d_slic-fts-clust-gc.html)

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 6 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: This will be needed for the inference usage you mentioned in another comment. But, I haven't been able to get it to work yet.

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 3 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: Thanks, I managed to format the things as required by flake8.

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 4 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: Done in next commit

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 5 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: Removed in next commit

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 1 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: Thanks, I managed to format the things as required by flake8.

**Comment by akshaykvnit** [comment]
@Borda @williamFalcon I have made most of the changes suggested by @Borda in the last review (with the exception of inference usage) and got the tests to pass. 

I haven't been able to implement the inference usage properly. That will take some more time. 

> @williamFalcon a bit more general thought is it would be better to write examples as ipython notebooks which we can transform to nice pages in Docs even with images as I did [here](https://pyimsegm.readthedocs.io/en/latest/notebooks/segment-2d_slic-fts-clust-gc.html)

I had initially implemented this in a Jupyter Notebook, but then the usage of `hparams` (using `argparse` and taking input from command line) won't be possible as they have been used in all your examples. So, if we need to convert the examples to iPython notebooks, then I could open a new PR, but do mention what can be done about the `hparams`.

**Comment by Borda** [comment]
> I haven't been able to implement the inference usage properly. That will take some more time.

do you want to do it here in this PR or would you prefer to make a follow-up PR?

> I had initially implemented this in a Jupyter Notebook, but then the usage of `hparams` (using `argparse` and taking input from command line) won't be possible as they have been used in all your examples. So, if we need to convert the examples to iPython notebooks, then I could open a new PR, but do mention what can be done about the `hparams`.

In this case, I have in mind to create a kind of lightning ZOO with the models and their particular trainers and then having notebook would be nice as an extra example with hparams can be simple derived without duplicating extra code... just building from prefabricated bricks... @williamFalcon  ?

**Comment by akshaykvnit** [comment]
> @akshaykvnit do you want to do it here in this PR or would you prefer to make a follow-up PR?

Yes, I'll make a follow-up PR.

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+import numpy as np
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torchvision
+import torchvision.transforms as transforms
+from torch.utils.data import DataLoader, Dataset
+from torchvision.models.segmentation import fcn_resnet50
+
+from PIL import Image
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: pls move PIL to the block above with "3rd-party packages"

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+import pytorch_lightning as pl
+
+
+class KITTI(Dataset):
+    def __init__(self, root_path, split='test', img_size=(1242, 376), transform=None):
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: pls add simple docstring what is this class about, in particular, explanation about the void and valid labels
personally I would pass the label indexes as a parameter in init with your listed indexes as default... this may simplify re-usage :]

File: pl_examples/full_examples/semantic_segmentation/semseg.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        Returns a list of absolute paths to images inside given `path`
+        '''
+        files_list = list()
+        for filename in os.listdir(path):
+            files_list.append(os.path.join(path, filename))
+        return files_list
+
+
+class SegModel(pl.LightningModule):
+    def __init__(self, hparams):
```
Comment (Borda) [inline pl_examples/full_examples/semantic_segmentation/semseg.py]: as examples are becoming part of the Docs, we shall add at least basic documentation with a brief description and reference to resources...

**Comment by Borda** [comment]
> > @akshaykvnit do you want to do it here in this PR or would you prefer to make a follow-up PR?
> 
> Yes, I'll make a follow-up PR.

btw, just curious how did you verify without inference? just see the validation/test curve?

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 9 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: I'm not sure what documentation you need here. Because none of the other examples have documentation for the `LightningModule` class.

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 8 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: I have added these things in the latest commit.

**Comment by akshaykvnit** [comment]
> > > @akshaykvnit do you want to do it here in this PR or would you prefer to make a follow-up PR?
> > 
> > 
> > Yes, I'll make a follow-up PR.
> 
> btw, just curious how did you verify without inference? just see the validation/test curve?

So, I implemented the reference using normal PyTorch and it works. You can [see the results here](https://github.com/akshaykvnit/pl-sem-seg/blob/master/pl_training.ipynb) in the last block of the notebook. I added the same code after `trainer.fit(model)` and it works properly. Thus, we can be sure that the training works.

However, I figured that you needed inference to work when `trainer.test(model)` is called. I haven't managed to properly implement in this fashion (some problem with the `test_step` and `test_end` functions). However, I shall try to fix that on my own (and open a new PR when done or open an issue if stuck).

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 7 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: Done in the latest commit.

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 9 above)
Comment (Borda) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: I ment someting like:
```
"""Image segmentation model.

This is a basic image segmentation model implemented with Lightning ...

References
- <link to source / publication>
```

**Comment by Borda** [comment]
@williamFalcon how about the docs for examples? I would say it shall contain them as it is part of Docs but probably it is not a blocker... :]

File: pl_examples/full_examples/semantic_segmentation/semseg.py (hunk 9 above)
Comment (akshaykvnit) [reply pl_examples/full_examples/semantic_segmentation/semseg.py]: Done in the latest commit

**Comment by williamFalcon** [comment]
@akshaykvnit awesome tutorial!

**Comment by williamFalcon** [comment]
@akshaykvnit 

```
 No module named 'torchvision.models.segmentation'
```

oops, looks like this test fails. Need to fix asap

**Comment by Borda** [comment]
Yeah I have noticed that almost all examples are not tested at all... We shall run them with a minimal data....
